src/operator/nn/group_norm.cc

```diff
@@ -0,0 +1,144 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * Copyright (c) 2019 by Contributors
+ * \file group_norm.cc
+ * \brief Implements Group Normalization (https://arxiv.org/abs/1803.08494).
+*/
+
+#include "group_norm-inl.h"
+#include <nnvm/op_attr_types.h>
+#include "../elemwise_op_common.h"
+
+namespace mxnet {
+namespace op {
+
+DMLC_REGISTER_PARAMETER(GroupNormParam);
+
+static bool GroupNormShape(const nnvm::NodeAttrs& attrs,
+                           mxnet::ShapeVector *in_shape,
+                           mxnet::ShapeVector *out_shape) {
+  const GroupNormParam& param = nnvm::get<GroupNormParam>(attrs.parsed);
+  using namespace mshadow;
+  CHECK_EQ(in_shape->size(), 3U) << "Input:[data, gamma, beta]";
+  const mxnet::TShape &dshape = in_shape->at(groupnorm::kData);
+  CHECK_GE(dshape.ndim(), 3U);
+  const int num_groups = param.num_groups;
+  CHECK_EQ(dshape[1] % num_groups, 0) << "# of channels must be divisible by # of groups";
+
+  if (!mxnet::ndim_is_known(dshape)) {
+    return false;
+  }
+
+  in_shape->at(groupnorm::kGamma) = mxnet::TShape(Shape1(num_groups));
+  in_shape->at(groupnorm::kBeta) = mxnet::TShape(Shape1(num_groups));
+
+  out_shape->clear();
+  out_shape->push_back(dshape);
+
+  mxnet::TShape moments_shape(2, 1);
+  moments_shape[0] = dshape[0];
+  moments_shape[1] = num_groups;
+  out_shape->push_back(moments_shape);
+  out_shape->push_back(moments_shape);
+  return true;
+}
+
+NNVM_REGISTER_OP(GroupNorm)
+.describe(R"code(Group normalization.
+
+Normalizes the channels of the input tensor by mean and variance, and applies a scale ``gamma`` as
+well as offset ``beta``.
+
+Assume the input has more than one dimension and we normalize along axis 1.
+We first compute the mean and variance along this axis and then
+compute the normalized output, which has the same shape as input, as following:
+
+.. math::
+
+  out = \frac{data - mean(data, axis)}{\sqrt{var(data, axis) + \epsilon}} * gamma + beta
```

Comment:
Change the docstring here to reflect the computation logic of group norm, i.e.,

```
data = data.reshape((N, num_groups, C // num_groups, ...))
out = \frac{data - mean(data, 2)}{\sqrt{var(data, 2) + \epsilon}} * gamma + beta
```

Reply:
Done.